reconcile/utils/acs_api.py

```diff
@@ -0,0 +1,328 @@
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/simpleaccessscopes/{id}
+        unrestricted = False
+        try:
+            check_len_attributes(["id", "name", "rules"], api_data)
+        except ValueError as e:
+            # it is valid for the default Unrestricted access scope to have null `rules`
+            if api_data.get("name") != "Unrestricted":
+                raise e
+            unrestricted = True
+
+        super().__init__(
+            id=api_data["id"],
+            name=api_data["name"],
+            clusters=[]
+            if unrestricted
+            else api_data["rules"].get("includedClusters", []),
+            namespaces=[]
+            if unrestricted
+            else api_data["rules"].get("includedNamespaces", []),
+            description=api_data.get("description", ""),
+        )
+
+
+class PermissionSet(BaseModel):
+    id: str
+    name: str
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/permissionsets/{id}
+        check_len_attributes(["id", "name"], api_data)
+
+        super().__init__(id=api_data["id"], name=api_data["name"])
+
+
+def check_len_attributes(attrs: list[Any], api_data: Any) -> None:
+    # generic attribute check function for expected types with valid len()
+    for attr in attrs:
+        value = api_data.get(attr)
+        if value is None or len(value) == 0:
+            raise ValueError(
+                f"Attribute '{attr}' must exist and not be empty\n\t{api_data}"
+            )
+
+
+class AcsApi:
+    def __init__(
+        self,
+        instance: Any,
+        timeout: int = 30,
+    ) -> None:
+        self.base_url = instance["url"]
+        self.token = instance["token"]
+        self.timeout = timeout
+        self.session = requests.Session()
+
+    def __enter__(self) -> Self:
+        return self
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        self.session.close()
+
+    def generic_request(
+        self, path: str, verb: str, json: Optional[Any] = None
+    ) -> requests.Response:
+        url = f"{self.base_url}{path}"
+        headers = {
+            "Authorization": f"Bearer {self.token}",
+            "Content-Type": "application/json",
+        }
+
+        if verb == "GET":
+            response = self.session.get(url, headers=headers, timeout=self.timeout)
+        elif verb == "DELETE":
+            response = self.session.delete(url, headers=headers, timeout=self.timeout)
+        elif verb == "POST":
+            response = self.session.put(
+                url, headers=headers, json=json, timeout=self.timeout
+            )
+        elif verb == "PUT":
+            response = self.session.post(
+                url, headers=headers, json=json, timeout=self.timeout
+            )
+        else:
+            raise ValueError(f"Unsupported request type: {verb}")
```

Comment:
`session` has a `request` method doing the same thing

```python
        response = self.session.request(
            verb,
            url,
            headers=headers,
            json=json,
            timeout=self.timeout,
        )
```

since this method is used internally in this class, skip verb validation is ok, or use typing to do the checking:

```python
Verb: TypeAlias = Literal["GET", "POST", "PUT", "DELETE"]

    def generic_request(
        self, path: str, verb: Verb, json: Optional[Any] = None
    ) -> requests.Response:
    ...
```

Reply:
[updated](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/26f6e065ac9c73596206f8b7e63a95cc0bf77dc0)